normandy/base/middleware.py

```diff
@@ -128,3 +136,122 @@ def process_request(self, request):
                 }
             )
         return response
+
+
+def default_username(email):
+    """Generate username for the Django user.
+
+    :arg str/unicode email: the email address to use to generate a username
+
+    :returns: str/unicode
+
+    """
+    # bluntly stolen from django-browserid
+    # store the username as a base64 encoded sha224 of the email address
+    # this protects against data leakage because usernames are often
+    # treated as public identifiers (so we can't use the email address).
+    return smart_text(base64.urlsafe_b64encode(
+        hashlib.sha1(force_bytes(email)).digest()
+    ).rstrip(b'='))
+
+
+class OIDCAccessTokenAuthenticationMiddleware:
+
+    def __init__(self, get_response):
+        self.get_response = get_response
+        # XXX Here we *could* do something like:
+        # if not settings.OIDC_ACCESS_TOKEN_ENABLED:
+        #    from django.core.exceptions import MiddlewareNotUsed
+        #    raise MiddlewareNotUsed
+        self.oidc_user_endpoint = settings.OIDC_USER_ENDPOINT
+        self.UserModel = auth.get_user_model()
+
+    def __call__(self, request):
+        response = self.process_request(request)
+        if not response:
+            response = self.get_response(request)
+        return response
+
+    def process_request(self, request):
+        matched = re.match(
+            # Note thet re.match requires it to match from the beginning
+            # otherwise it might match `Something Bearer deadbeef`
+            r'Bearer\s+(.*)',
+            request.META.get('HTTP_AUTHORIZATION') or '',
+            re.I
+        )
+        if not matched:
+            # That's cool. If it's not there, it's not there.
+            # This middleware will not be able to authenticate the user
+            # for the duration of this request.
+            return
+        access_token = matched.group(1)
+        cache_key = 'oidc-access-token-check:{}'.format(
+            hashlib.md5(force_bytes(access_token)).hexdigest()
+        )
+        checked = cache.get(cache_key)
+        if checked:
+            return checked
+        # print("access_token", repr(access_token))
+        # XXX This should use a retrying requests sesssion or backoff lib.
+        response = requests.get(self.oidc_user_endpoint, headers={
+            'Authorization': 'Bearer {0}'.format(access_token)
+        })
+        if response.status_code == 200:
+            # Awesome! Turn this into a Django User
+            user_profile = response.json()
+            if 'email' not in user_profile:
+                raise ImproperlyConfigured(
+                    "The HTTP Authorization header had an access token that "
+                    "was created without the 'email' scope."
+                )
+            email = user_profile['email']
+            if not email:
```

Comment:
if ther ewasn't an `email` scope, wouldn't the `if 'email' not in user_profile` catch it first?

Reply:
No, because if the email is `''` (the empty string) then it is `in user_profile`, but is not truthy, and so the code will take the `if not email:` branch.

Reply:
I think I'm going to collapse these and only use the `PermissionDenied`.

The verbose comment below is still useful but now it's actually situated in the wrong place. 
If someone does, correctly, use the `email` scope and super-unlikely come with a blank email then we simply don't want to let them in. Perhaps there's something in the RFC of OAuth2.0 about the email having to be a real email if you set the scope but I'm too lazy to dig into that. No email, permission denied. Using `ImproperlyConfigured` was wrong.